dockstore-client/src/main/java/io/dockstore/client/cli/nested/BaseLauncher.java

```diff
@@ -0,0 +1,103 @@
+package io.dockstore.client.cli.nested;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import io.dockstore.common.FileProvisioning;
+import io.dockstore.common.LanguageType;
+import org.apache.commons.io.FileUtils;
+
+/**
+ * This class is the base class for launchers used by the Dockstore CLI.
+ * Launchers such as cwltool and cromwell extend this.
+ */
+public abstract class BaseLauncher {
+    protected final AbstractEntryClient abstractEntryClient;
+    protected final FileProvisioning fileProvisioning;
+    // The primary descriptor of the workflow
+    protected File primaryDescriptor;
+    // A zip file for the entire entry
+    protected File importsZip;
+    // Parameter file with all remote files resolved locally
+    protected File provisionedParameterFile;
+    // Path to original parameter file
+    protected String originalParameterFile;
+    // The working directory
+    // For local entries this is the parent dir of the primary file
+    // For remote entries this is the tmp dir where workflow files are downloaded to
+    protected String workingDirectory;
+
+    // CWL, WDL, NEXTFLOW
+    protected LanguageType languageType;
+    protected File exectionFile;
+
+    protected boolean script;
+
+    public BaseLauncher(AbstractEntryClient abstractEntryClient, LanguageType language, boolean script) {
+        this.abstractEntryClient = abstractEntryClient;
+        this.fileProvisioning = new FileProvisioning(abstractEntryClient.getConfigFile());
+        this.languageType = language;
+        this.script = script;
+    }
+
+    /**
+     * Set settings for launcher relevant to the current run
+     * @param descriptor
+     * @param imports
+     * @param provisionedParameters
+     * @param originalParameters
+     * @param workDir
+     */
+    public void setFiles(File descriptor, File imports, File provisionedParameters, String originalParameters, String workDir) {
+        this.primaryDescriptor = descriptor;
+        this.importsZip = imports;
+        this.provisionedParameterFile = provisionedParameters;
+        this.originalParameterFile = originalParameters;
+        this.workingDirectory = workDir;
+    }
+
+    /**
+     * Download and install the launcher if necessary
+     */
+    public abstract void initialize();
+
+    /**
+     * Create a command to execute entry on the command line
+     * @return Command to run entry
+     */
+    public abstract String buildRunCommand();
```

Comment:
One thought is that this should return a `List<String>`. There have been potential, maybe only hypothetical issues, where if an argument has an embedded space it needs to be escaped. If this returns a list of arguments instead then escaping could be handled at the very end, instead of each implementation subclass needing to handle it.